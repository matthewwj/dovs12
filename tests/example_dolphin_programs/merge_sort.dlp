void printInt(i : int){
   output_string(int_to_string(i),get_stdout());
   output_string("\n",get_stdout());
}

record Singly { head: int; next: Singly; }

Singly split(l: Singly) {
    var n = 0;
    var l1 = l;
    while (l1 != nil) { l1 = l1.next; n = n+1; };
    n = (n-1)/2;
    var last = l;
    while (n > 0) { last = last.next; n = n-1; };
    var l2 = last.next;
    last.next = nil;
    return l2;
}

Singly merge(a: Singly, b: Singly) {
    if (a == nil) return b;
    if (b == nil) return a;
    if (a.head > b.head) return merge(b, a);
    var res = a; // pointer to the very first element
    var l1: Singly = a;
    var l2: Singly = b;
    // invariant res->...->l1->...->null
    //                     l2->...->null
    //       and l1.head <= l2.head
    while (l2 != nil) {
        var v = l2.head;
        var n = l1.next;
        while (n != nil && n.head <= v) { l1 = n; n = n.next; };
        l1.next = l2;
        l2 = n;
    };
    return res;
}

Singly mergesort(l: Singly) {
    if (l == nil || l.next == nil) return l;
    var l2 = split(l);
    return merge(mergesort(l), mergesort(l2));
}

void print_list(l: Singly) {
    if (l == nil) { output_string("\n", get_stdout()); return; };
    printInt(l.head); 
    print_list(l.next);
}

record state { s : int; }
int random(s : state) { s.s = (s.s * 987 + 42) % 5003; return s.s; }

Singly random_list(n: int, s : state) {
    if (n == 0) return nil;
    return new Singly { head = random(s); next = random_list(n - 1, s);};
}

int main() {
    var l = new Singly { head = 4; next = new Singly { head = 1; next = new Singly { head = 2; next = new Singly { head = 5 ; next = new Singly { head = 3; next = nil;};};};};};
    print_list(l);
    print_list(mergesort(l));
    var l2 = random_list(100, new state { s = 0; });
    print_list(l2);
    print_list(mergesort(l2));
    return 0;
}