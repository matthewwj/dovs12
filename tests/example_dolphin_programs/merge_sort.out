4
1
2
5
3

1
2
3
4
5

42
1472
2036
3371
224
998
4480
4153
1596
4352
2892
2736
3857
4621
3236
2060
2044
1261
3905
1967
307
2871
2021
3575
1452
2308
1673
303
3926
2682
589
1037
2949
3962
3193
4646
2896
1681
3196
2604
3651
1419
4758
3374
3185
1753
4218
712
2366
3886
3226
2196
1195
3802
366
1068
3528
90
3821
4110
4182
201
3312
2027
4494
2962
1784
4797
1843
2994
3350
4512
716
1311
3225
1209
2611
554
1513
2479
348
3314
4001
1662
4455
4493
1975
3200
1549
2990
4405
170
2733
896
3866
3498
498
1274
1727
3571

42
90
170
201
224
303
307
348
366
498
554
589
712
716
896
998
1037
1068
1195
1209
1261
1274
1311
1419
1452
1472
1513
1549
1596
1662
1673
1681
1727
1753
1784
1843
1967
1975
2021
2027
2036
2044
2060
2196
2308
2366
2479
2604
2611
2682
2733
2736
2871
2892
2896
2949
2962
2990
2994
3185
3193
3196
3200
3225
3226
3236
3312
3314
3350
3371
3374
3498
3528
3571
3575
3651
3802
3821
3857
3866
3886
3905
3926
3962
4001
4110
4153
4182
4218
4352
4405
4455
4480
4493
4494
4512
4621
4646
4758
4797

0
